File: java/src/leetcode/P0070.java
package leetcode;

public class P0070 {
    public static void main(String[] args) {
        System.out.println(climbStairs(1));
    }
/*
    D[1] = 1
    D[2] = D[1] + D[0]
    D[3] = D[2] + D[1]
    D[4] = D[3] + D[2]
    1111
    112
    121
    211
    22
    ...
    D[N] = D[N-1] + 1
*/

    public static int climbStairs(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        int[] memo = new int[n];
        memo[0] = 1;
        memo[1] = 2;
        for (int i = 2; i < n; i++) {
            memo[i] = memo[i - 1] + memo[i - 2];
        }

        return memo[n-1];
    }
}

File: ruby/leetcode/problems/P0070.rb
# @param {Integer} n
# @return {Integer}
def climb_stairs(n)
  # dp(0) = 0
  # dp(1) = 1
  # dp(2) = 2
  # dp(3) = dp(1) + 2 # 111, 12, 21 = dp(2) + 1 = dp(2) + dp(1)
  # dp(4) = dp(2) + 2 # 1111, 112, 211, 121, 22 =  dp(3) +dp(2)
  memo = [46]
  memo[0] = 0
  memo[1] = 1
  memo[2] = 2
  (3..n).each { |i|
    memo[i] = memo[i - 1] + memo[i - 2]
  }
  memo[n]
end

n = 4
puts climb_stairs(n)